pkg/compactor/shuffle_sharding_grouper.go

```diff
@@ -142,103 +162,164 @@ func (g *ShuffleShardingGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (re
-				g.hashFunc,
-				g.blockFilesConcurrency,
-				g.blocksFetchConcurrency,
-			)
-			if err != nil {
-				return nil, errors.Wrap(err, "create compaction group")
-			}
-
-			for _, m := range group.blocks {
-				if err := thanosGroup.AppendMeta(m); err != nil {
-					return nil, errors.Wrap(err, "add block to compaction group")
-				}
-			}
-
-			outGroups = append(outGroups, thanosGroup)
-		}
+		groups = append(groups, groupBlocksByCompactableRanges(mainBlocks, g.compactorCfg.BlockRanges.ToMilliseconds())...)
 	}
 
 	// Ensure groups are sorted by smallest range, oldest min time first. The rationale
 	// is that we wanna favor smaller ranges first (ie. to deduplicate samples sooner
 	// than later) and older ones are more likely to be "complete" (no missing block still
 	// to be uploaded).
-	sort.SliceStable(outGroups, func(i, j int) bool {
-		iLength := outGroups[i].MaxTime() - outGroups[i].MinTime()
-		jLength := outGroups[j].MaxTime() - outGroups[j].MinTime()
+	sort.SliceStable(groups, func(i, j int) bool {
+		iGroup := groups[i]
+		jGroup := groups[j]
+		iMinTime := iGroup.minTime()
+		iMaxTime := iGroup.maxTime()
+		jMinTime := jGroup.minTime()
+		jMaxTime := jGroup.maxTime()
+		iLength := iMaxTime - iMinTime
+		jLength := jMaxTime - jMinTime
 
 		if iLength != jLength {
 			return iLength < jLength
 		}
-		if outGroups[i].MinTime() != outGroups[j].MinTime() {
-			return outGroups[i].MinTime() < outGroups[j].MinTime()
+		if iMinTime != jMinTime {
+			return iMinTime < jMinTime
 		}
 
+		iGroupHash := hashGroup(g.userID, iGroup.rangeStart, iGroup.rangeEnd)
+		iGroupKey := createGroupKey(iGroupHash, iGroup)
+		jGroupHash := hashGroup(g.userID, jGroup.rangeStart, jGroup.rangeEnd)
+		jGroupKey := createGroupKey(jGroupHash, jGroup)
 		// Guarantee stable sort for tests.
-		return outGroups[i].Key() < outGroups[j].Key()
+		return iGroupKey < jGroupKey
 	})
 
+mainLoop:
+	for _, group := range groups {
+		var blockIds []string
+		for _, block := range group.blocks {
+			blockIds = append(blockIds, block.ULID.String())
+		}
+		blocksInfo := strings.Join(blockIds, ",")
+		level.Info(g.logger).Log("msg", "check group", "blocks", blocksInfo)
+
+		// Nothing to do if we don't have at least 2 blocks.
+		if len(group.blocks) < 2 {
+			continue
+		}
+
+		groupHash := hashGroup(g.userID, group.rangeStart, group.rangeEnd)
+
+		if isLocked, err := g.isGroupLocked(group.blocks); err != nil {
+			level.Warn(g.logger).Log("msg", "unable to check if blocks in group are locked", "group hash", groupHash, "err", err, "group", group.String())
+			continue
+		} else if isLocked {
+			level.Info(g.logger).Log("msg", "skipping group because at least one block in group are locked", "group_hash", groupHash)
+			continue
+		}
+
+		remainingCompactions++
+		groupKey := createGroupKey(groupHash, group)
+
+		level.Info(g.logger).Log("msg", "found compactable group for user", "group_hash", groupHash, "group", group.String())
+		g.lockBlocks(group.blocks)
```

Comment:
Moved goroutine periodically updating lock into planner. So the case would be like:

1. Compactor-1 - grouper sees blocks aren't locked
2. Compactor-2 - grouper sees blocks aren't locked
3. Compactor-1 - grouper locks the block
4. Compactor-2 - grouper locks the block
5. Compactor-1 - planner checks blocks are locked by Compactor-1. failed and stop compaction
6. Compactor-2 - planner checks blocks are locked by Compactor-2
7. Compactor-2 - planner kicks off goroutine that periodically updates the lock

This logic should minimize the case two compactors trying to compact with same block. The case it would happen is that Compactor-2 locks the first block after Compactor-1 planner starts checking blocks. This would be very hard to happen since bucket on Compactor-2 would be super slow that it did not finish upsert one lock file while Compactor-1 finishes upsert all lock files.